### PR TITLE
Add Shader.Find() support for YAML shaders

### PR DESCRIPTION
**Pull request by trouger**
https://github.com/trouger/AssetRipper/issues/9

### CODE REVIEW

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    public class AssetRipperShaderPostprocessor : AssetPostprocessor
+    {
+        static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAssets, string[] movedAssets, string[] movedFromAssetPaths)
+        {
+            foreach (var importedAsset in importedAssets)
+            {
+                if (!importedAsset.EndsWith(".asset")) continue;
+                Shader yamlShader = AssetDatabase.LoadMainAssetAtPath(importedAsset) as Shader;
+                if (yamlShader == null) continue;
+                ShaderUtil.RegisterShader(yamlShader);
```
Comment (ds5678) [inline AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:20]: Is this function forgiving?
* What happens if this is used multiple times on the same shader?
* What about a normal shader? If you didn't have a file extension check, would that break things?

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    /// Such a shader can be assigned to and used by a material as a regular .shader asset,
+    /// but it does not work with Shader.Find() unless we explicitly register it. 
+    /// </summary>
+    public class AssetRipperShaderPostprocessor : AssetPostprocessor
+    {
+        static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAssets, string[] movedAssets, string[] movedFromAssetPaths)
+        {
+            foreach (var importedAsset in importedAssets)
+            {
+                if (!importedAsset.EndsWith(".asset")) continue;
```
Comment (ds5678) [inline AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:17]: If Unity allows Ordinal string comparisons, that method overload should be used.

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+using UnityEngine;
+using UnityEditor;
+
+namespace AssetRipperPatches.Editor
+{
+    /// <summary>
+    /// This script is AssetRipper's patch for shaders exported as YAML assets.
+    /// Such a shader can be assigned to and used by a material as a regular .shader asset,
+    /// but it does not work with Shader.Find() unless we explicitly register it.
```
Comment (ds5678) [inline AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:9]: I saw you document on the other file that this is only applicable to 2018 +. I would also like it mentioned here.

File: AssetRipper.Library/Exporters/Shaders/YamlShaderExportCollection.cs [hunk 4]
```diff
@@ -10,5 +12,17 @@ public YamlShaderExportCollection(YamlShaderExporter assetExporter, IUnityObject
 		}
 
 		protected override string GetExportExtension(IUnityObjectBase asset) => "asset";
+		
+		protected override bool ExportInner(IProjectAssetContainer container, string filePath, string dirPath)
+		{
+			// This patch uses ShaderUtil.RegisterShader(), which is only available start from Unity 2018.
```
Comment (ds5678) [inline AssetRipper.Library/Exporters/Shaders/YamlShaderExportCollection.cs:18]: > This patch uses ShaderUtil.RegisterShader()

I appreciate the specificity here. Thank you.

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt (hunk 2 above)
Comment (trouger) [reply AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:17]: Yes, even Unity 5 has string comparison overloads. I think OrdinalIgnoreCase should be used here.

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt (hunk 1 above)
Comment (trouger) [reply AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:20]: What happens if this is used multiple times on the same shader?
Totally allowed. Same effect as a single time.

What about a normal shader? If you didn't have a file extension check, would that break things?
A normal shader is automatically registered. Registering it again by using ShaderUtil.RegisterShader is much the same as the situation in the first question - registering multiple times.
The file extension check is solely for lowering the performance cost, as we don't want to load all meshes and textures in the project.

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt (hunk 2 above)
Comment (ds5678) [reply AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt]: Not ignore case, just Ordinal

File: AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt (hunk 1 above)
Comment (ds5678) [reply AssetRipper.Library/Exporters/Shaders/UnityPatch/YamlShaderPostprocessor.txt:23]: Perfect

**Review by ds5678 — CHANGES_REQUESTED** [review]
Just switch `OrdinalIgnoreCase` for `Ordinal` and this is good for merging!

**Review by ds5678 — APPROVED** [review]
Looks fantastic. Thank you for contributing!

**Pull request merged**